pkg/oidc/jwks.go

```diff
@@ -2,61 +2,56 @@ package oidc
 
 import (
 	"bytes"
+	"crypto"
 	"crypto/rsa"
-	"crypto/sha256"
 	"crypto/x509"
 	"encoding/base64"
 	"encoding/json"
 	"fmt"
-	"math/big"
+
+	jose "gopkg.in/square/go-jose.v2"
 )
 
 type KeyResponse struct {
-	Keys []*Keys `json:"keys"`
+	Keys []jose.JSONWebKey `json:"keys"`
 }
 
-type Keys struct {
-	Kty string `json:"kty"`
-	Alg string `json:"alg"`
-	Use string `json:"use"`
-	Kid string `json:"kid"`
-	N   string `json:"n"`
-	E   string `json:"e"`
-}
+// copied from kubernetes/kubernetes#78502
+func digestOfKey(key *rsa.PrivateKey) (string, error) {
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(&key.PublicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize public key to DER format: %v", err)
+	}
+
+	hasher := crypto.SHA256.New()
+	hasher.Write(publicKeyDERBytes)
+	publicKeyDERHash := hasher.Sum(nil)
 
-func digestOfKey(key *rsa.PrivateKey) string {
-	s := sha256.New()
-	s.Write(x509.MarshalPKCS1PrivateKey(key))
-	return base64.RawURLEncoding.EncodeToString(s.Sum(nil))
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	return keyID, nil
 }
 
 func GenerateKeysFile(key *rsa.PrivateKey) (*bytes.Reader, error) {
-	keyE := base64.RawURLEncoding.EncodeToString(big.NewInt(int64(key.E)).Bytes())
-	keyN := base64.RawURLEncoding.EncodeToString(key.N.Bytes())
-	v := KeyResponse{
-		Keys: []*Keys{
-			{
-				Kty: "RSA",
-				Alg: "RS256",
-				Use: "sig",
-				Kid: digestOfKey(key),
-				E:   keyE,
-				N:   keyN,
-			},
-			{
-				Kty: "RSA",
-				Alg: "RS256",
-				Use: "sig",
```

Comment:
we can drop that part and simplify it, empty kid was used before K8s 1.16 and it's not relevant for us